async_simple/coro/ConditionVariable.h

```diff
@@ -0,0 +1,200 @@
+        : _cv(cv), _lock(lock) {}
+
+    bool await_ready() const noexcept { return false; }
+
+    bool await_suspend(std::coroutine_handle<> continuation) noexcept {
+        _continuation = continuation;
+        std::unique_lock<Lock> lock{_lock, std::adopt_lock};
+        auto awaitings = _cv->_awaiters.load(std::memory_order_relaxed);
+        do {
+            _next = awaitings;
+        } while (!_cv->_awaiters.compare_exchange_weak(
+            awaitings, this, std::memory_order_acquire,
+            std::memory_order_relaxed));
+        return true;
+    }
+    void await_resume() const noexcept {}
+
+public:
+    ConditionVariable<Lock>* _cv;
+    Lock& _lock;
+
+private:
+    friend class ConditionVariable<Lock>;
+    ConditionVariableAwaiter<Lock>* _next = nullptr;
+    std::coroutine_handle<> _continuation;
+};
+
+template <class Lock>
+template <class Pred>
+inline Lazy<> ConditionVariable<Lock>::wait(Lock& lock, Pred&& pred) noexcept {
+    while (!pred()) {
+        co_await ConditionVariableAwaiter<Lock>{this, lock};
+        co_await lock.coLock();
+    }
+    co_return;
+}
+
+template <class Lock>
+inline void ConditionVariable<Lock>::notify() noexcept {
+    auto awaitings = _awaiters.load(std::memory_order_relaxed);
+    while (!_awaiters.compare_exchange_weak(awaitings, nullptr,
+                                            std::memory_order_release,
+                                            std::memory_order_relaxed))
+        ;
+    resumeWaiters(awaitings);
+}
+
+template <class Lock>
+inline void ConditionVariable<Lock>::resumeWaiters(
+    ConditionVariableAwaiter<Lock>* awaiters) {
+    for (auto head = awaiters; head; head = head->_next) {
+        head->_continuation.resume();
+    }
+}
+
+template <>
+class ConditionVariableAwaiter<>;
+template <>
+class ConditionVariable<> {
+public:
+    using pointer_type = void*;
+
+public:
+    ConditionVariable() noexcept {}
+    ~ConditionVariable() {}
+
+    ConditionVariable(const ConditionVariable&) = delete;
+    ConditionVariable& operator=(const ConditionVariable&) = delete;
+
+    void notify() noexcept;
+    ConditionVariableAwaiter<> wait() noexcept;
+    void reset() noexcept;
+
+private:
+    void resumeWaiters(ConditionVariableAwaiter<>* awaiters);
+
+private:
+    friend class ConditionVariableAwaiter<>;
+    std::atomic<pointer_type> _awaiters = nullptr;
+};
+
+template <>
+class ConditionVariableAwaiter<> {
+public:
+    using pointer_type = void*;
+
+public:
+    ConditionVariableAwaiter(ConditionVariable<>* cv) noexcept : _cv(cv) {}
+
+    bool await_ready() const noexcept {
+        return static_cast<pointer_type>(_cv) ==
+               _cv->_awaiters.load(std::memory_order_acquire);
+    }
+    bool await_suspend(std::coroutine_handle<> continuation) noexcept {
+        _continuation = continuation;
+        pointer_type awaitings = _cv->_awaiters.load(std::memory_order_acquire);
+        do {
```

Comment:
```suggestion
        pointer_type awaitings = nullptr;
        do {
             awaiting =  _cv->_awaiters.load(std::memory_order_acquire);
```

Otherwise, the `awaitings` wouldn't be updated in case the condition in while fails.

Reply:
the `awaitings` will be updated to `_cv->_awaiters`'s latest value automatically when `cas` failed

Reply:
Oh, you are correct here.